cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/MultipleOutputs.java

```diff
@@ -0,0 +1,280 @@
+
+        Instantiator<? extends OutputFormat<?, ?>> instantiator =
+          new InstantiatorFactory(false).get(TypeToken.of(taskContext.getOutputFormatClass()));
+        OutputFormat<?, ?> outputFormat = instantiator.create();
+
+        writer = outputFormat.getRecordWriter(taskContext);
+      } catch (ClassNotFoundException e) {
+        throw new IOException(e);
+      }
+
+      // add the record-writer to the cache
+      recordWriters.put(namedOutput, writer);
+    }
+    return writer;
+  }
+
+  // Create a taskAttemptContext for the named output with
+  // output format and output key/value types put in the context
+  private TaskAttemptContext getContext(String nameOutput) throws IOException {
+
+    TaskAttemptContext taskContext = taskContexts.get(nameOutput);
+
+    if (taskContext != null) {
+      return taskContext;
+    }
+
+    taskContext = getNamedContext(context, nameOutput);
+    taskContexts.put(nameOutput, taskContext);
+    return taskContext;
+  }
+
+  static TaskAttemptContext getNamedContext(TaskAttemptContext context, String nameOutput) throws IOException {
+    // The following trick leverages the instantiation of a record writer via
+    // the job thus supporting arbitrary output formats.
+    Job job = new Job(context.getConfiguration());
+    job.setOutputFormatClass(getNamedOutputFormatClass(context, nameOutput));
+    job.setOutputKeyClass(getNamedOutputKeyClass(context, nameOutput));
+    job.setOutputValueClass(getNamedOutputValueClass(context, nameOutput));
+
+    Configuration conf = job.getConfiguration();
+    Map<String, String> namedConfigurations = getNamedConfigurations(context, nameOutput);
+    for (Map.Entry<String, String> entry : namedConfigurations.entrySet()) {
+      conf.set(entry.getKey(), entry.getValue());
+    }
+    return new TaskAttemptContextImpl(conf, context.getTaskAttemptID(), new WrappedStatusReporter(context));
+  }
+
+  private static class WrappedStatusReporter extends StatusReporter {
+
+    TaskAttemptContext context;
+
+    public WrappedStatusReporter(TaskAttemptContext context) {
+      this.context = context;
+    }
+
+    @Override
+    public Counter getCounter(Enum<?> name) {
+      return context.getCounter(name);
+    }
+
+    @Override
+    public Counter getCounter(String group, String name) {
+      return context.getCounter(group, name);
+    }
+
+    @Override
+    public void progress() {
+      context.progress();
+    }
+
+    @Override
+    public float getProgress() {
+      return context.getProgress();
+    }
+
+    @Override
+    public void setStatus(String status) {
+      context.setStatus(status);
+    }
+  }
+
+  /**
+   * Closes all the opened outputs.
+   * This should be called from cleanup method of map/reduce task.
+   */
+  @SuppressWarnings("unchecked")
+  public void close() throws IOException, InterruptedException {
+    for (RecordWriter writer : recordWriters.values()) {
+      writer.close(context);
```

Comment:
Is it suggested to LOG.ERROR if one fails? or try to close all of them and then propagate the exception at the end?